pkg/agent/client.go

```diff
@@ -76,33 +85,154 @@ type AgentClient struct {
+
+	// connect opts
+	address string
+	opts    []grpc.DialOption
+	conn    *grpc.ClientConn
+	stopCh  chan struct{}
+	// locks
+	sendLock      sync.Mutex
+	recvLock      sync.Mutex
+	probeInterval time.Duration // interval between probe pings
+
+	// file path contains service account token.
+	// token's value is auto-rotated by kubernetes, based on projected volume configuration.
+	serviceAccountTokenPath string
 }
 
 func newAgentClient(address, agentID string, cs *ClientSet, opts ...grpc.DialOption) (*AgentClient, error) {
-	stream, err := NewRedialableAgentClient(address, agentID, cs, opts...)
+	a := &AgentClient{
+		cs:                      cs,
+		address:                 address,
+		agentID:                 agentID,
+		opts:                    opts,
+		probeInterval:           cs.probeInterval,
+		stopCh:                  make(chan struct{}),
+		serviceAccountTokenPath: cs.serviceAccountTokenPath,
+		connManager:             newConnectionManager(),
+	}
+	err := a.Connect()
 	if err != nil {
 		return nil, err
 	}
-	return newAgentClientWithRedialableAgentClient(stream), nil
+	return a, nil
 }
 
-func newAgentClientWithRedialableAgentClient(rac *RedialableAgentClient) *AgentClient {
-	return &AgentClient{
-		connManager: newConnectionManager(),
-		stream:      rac,
-		stopCh:      rac.stopCh,
+// Connect makes the grpc dial to the proxy server. It returns the serverID
+// it connects to.
+func (a *AgentClient) Connect() error {
+	conn, err := grpc.Dial(a.address, a.opts...)
+	if err != nil {
+		return err
+	}
+	ctx := metadata.AppendToOutgoingContext(context.Background(), header.AgentID, a.agentID)
+	if a.serviceAccountTokenPath != "" {
+		if ctx, err = a.initializeAuthContext(ctx); err != nil {
+			conn.Close()
+			return err
+		}
+	}
+	stream, err := agent.NewAgentServiceClient(conn).Connect(ctx)
+	if err != nil {
+		conn.Close()
+		return err
+	}
+	serverID, err := serverID(stream)
+	if err != nil {
+		conn.Close()
+		return err
+	}
+	serverCount, err := serverCount(stream)
+	if err != nil {
+		conn.Close()
+		return err
 	}
+	a.conn = conn
+	a.stream = stream
+	a.serverID = serverID
+	klog.Infof("Connect to server %s", serverID)
+	a.serverCount = serverCount
+	return nil
 }
 
-// Close closes the underlying stream.
+// Close closes the underlying connection.
 func (a *AgentClient) Close() {
-	if a.stream == nil {
-		klog.Warning("Unexpected empty AgentClient.stream")
-		return
+	if a.conn == nil {
+		klog.Warning("Unexpected empty AgentClient.conn")
 	}
-	a.stream.Close()
+	a.conn.Close()
+	close(a.stopCh)
+}
+
+func (a *AgentClient) Send(pkt *client.Packet) error {
+	a.sendLock.Lock()
+	defer a.sendLock.Unlock()
+
+	err := a.stream.Send(pkt)
+	// TODO: add metrics for send err
+	if err != nil && err != io.EOF {
+		a.cs.RemoveClient(a.serverID)
+	}
+	return err
+}
+
+func (a *AgentClient) Recv() (*client.Packet, error) {
+	a.recvLock.Lock()
+	defer a.recvLock.Unlock()
+
+	pkt, err := a.stream.Recv()
```

Comment:
Not blocking for  this PR but a little worried that Recv is blocking and may stop us shutting down properly.

Reply:
Good point. I did some research on this and it seems fine.

When we close the agent client, we close the underlying GRPC connection, which should close the stream and unblocks Recv() with an error, according to https://github.com/grpc/grpc-go/issues/3230#issuecomment-562242727.